File: 2448-count-number-of-bad-pairs/count-number-of-bad-pairs.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public long countBadPairs(int[] nums){
        int n = nums.length;
        long totalPairs = (long) n * (n - 1) / 2;

        Map<Integer, Integer> diffs = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int diff = nums[i] - i;
            diffs.put(diff, diffs.getOrDefault(diff, 0) + 1);
        }

        for (Map.Entry<Integer, Integer> entry : diffs.entrySet()) {
            int v = entry.getValue();
            totalPairs -= (long) v * (v - 1) / 2;
        }
        return totalPairs;
    }
}

File: 2448-count-number-of-bad-pairs/count-number-of-bad-pairs.py
'''
    4   1   3   3
    0   1   2   3
    4   0   1   0

1
6 - 1

3 * 2 - 1

0 - 2
4 - 1
1 - 1

'''

class Solution:
    def countBadPairs(self, nums: List[int]) -> int:
        diffs = collections.defaultdict(int)
        for i, num in enumerate(nums):
            diffs[num - i] += 1

        n = len(nums)
        total = n * (n - 1) // 2
        for d in diffs.values():
            total -= d * (d - 1) // 2

        return total